main.cpp

```diff
@@ -442,97 +429,110 @@ struct ip_set
+		out->sock_idx = ent.sock_idx;
 
 		return true;
 	}
 
-	void insert_node(udp::endpoint const& ep, char const* node_id
-		, bound_socket& incoming_socket)
+	void insert_node(Address const& addr, std::uint16_t const port, int const sock_idx)
 	{
-		assert(ep.address() != address_v4::any());
-
-		if (ep.protocol() == udp::v4()
-			&& m_ip4s.count(ep.address().to_v4()))
-			return;
-		else if (ep.protocol() == udp::v6()
-			&& m_ip6s.count(ep.address().to_v6()))
-			return;
-
-		// don't let the queue get too big
-		if (m_queue.size() > ping_queue_size) return;
-
-		// as the size approaches the limit, increasingly reject
-		// new nodes, to distribute nodes we ping more evenly
-		// over time
-		++m_round_robin;
-		m_round_robin &= 0xff;
-		if (m_round_robin < m_queue.size() * 256 / ping_queue_size)
-			return;
+		assert(addr != Address::any());
+
+		// prevent duplicate entries
+		if (m_ips.count(addr)) return;
+
+		// the number of nodes we allow in the queue before we start dropping
+		// nodes (to stay below the limi)
+		size_t const low_watermark = ping_queue_size / 2;
+
+		if (m_queue.size() > low_watermark) {
+			// as the size approaches the limit, increasingly reject
+			// new nodes, to distribute nodes we ping more evenly
+			// over time.
+			// we don't start dropping nodes until we exceed the low watermark, but
+			// once we do, we increase the drop rate the closer we get to the limit
+			++m_round_robin;
+			if (m_round_robin < (m_queue.size() - low_watermark) * 256 / (ping_queue_size - low_watermark))
+				return;
+		}
 
-		queued_node_t e;
-		e.ep = ep;
-		memcpy(e.node_id.data(), node_id, e.node_id.size());
-		e.incoming_socket = &incoming_socket;
 		// we primarily want to keep quality nodes in our list.
 		// in 10 minutes, any pin-hole the node may have had open to
 		// us is likely to have been closed. If the node responds
-		// in 10 minutes from now, it's likely to either have a full-cone
+		// in 15 minutes from now, it's likely to either have a full-cone
 		// NAT or not be NATed at all (which is the way we like our nodes).
 		// also, it still being up is a good predictor for it staying up
 		// longer as well.
-		e.expire = steady_clock::now() + minutes(15);
+		std::uint32_t const expire = duration_cast<seconds>(
+			steady_clock::now() + minutes(15) - m_created).count();
 
-		m_queue.push_back(e);
-		if (ep.protocol() == udp::v4())
-			m_ip4s.insert(ep.address().to_v4());
-		else
-			m_ip6s.insert(ep.address().to_v6());
+		m_queue.push_back({addr.to_bytes(), std::uint32_t(sock_idx), expire, port});
+		m_ips.insert(addr);
 	}
 
 private:
 
+	// this is the type of each node queued up to be pinged at some point in the
+	// future
+	struct queue_entry_t
+	{
+		typename Address::bytes_type addr;
+
+		// the index of the socket this node was seen on, and should be pinged
+		// back via
+		std::uint32_t sock_idx:2;
+
+		// expiration in seconds (relative to the ping_queue creation time)
+		std::uint32_t expire:30;
```

Comment:
I suppose technically there will always be  bytes padding here, so the bitfield may be a bit redundant